From b9e8085f15bd31684f8cc7b56b1cb99af697121f Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Wed, 6 Nov 2024 13:57:06 +0100
Subject: [PATCH 1355/1695] drivers: stm32_rtc: fix exti wakeup management

The stm32_exti_enable_wake/stm32_exti_disable_wake can be called
directly in stm32_rtc_set_alarm_wakeup_status() as the EXTI already
manage the wake up interruption for low power mode.

This patch allows to simplify the code, avoids unneeded RTC PM function,
with access to EXTI and avoids potential issue if stm32_rtc_pm() callback
is called after EXTI pm function.

This patch also replaces the flags "STM32_EXTI" by "CFG_STM32_EXTI".

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6a3f7c597d6d5b132bf54938bbf90d58e2ebb2fa drivers: stm32_rtc: fix exti wakeup management]
Fixes: 01bb4acfbb54 ("drivers: stm32_rtc: add the capability to wakeup the platform")
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I27c2ec12942cb4f1da74d07c325b7d2ffb4c87ca
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/416821
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Clement LE GOFFIC <clement.legoffic@foss.st.com>
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
---
 core/drivers/stm32_rtc.c | 35 ++++++++++++-----------------------
 1 file changed, 12 insertions(+), 23 deletions(-)

diff --git a/core/drivers/stm32_rtc.c b/core/drivers/stm32_rtc.c
index 02ee6a8d7..f8ea664c5 100644
--- a/core/drivers/stm32_rtc.c
+++ b/core/drivers/stm32_rtc.c
@@ -207,6 +207,8 @@ struct rtc_compat {
  * @itr_handler: Interrupt handler
  * @notif_id: Notification ID
  * @wait_alarm_return_status: Status of the wait alarm
+ * @exti: EXTI data associated to wake-up interruption
+ * @rtc: information for OP-TEE RTC device
  */
 struct rtc_device {
 	struct io_pa_va base;
@@ -224,7 +226,6 @@ struct rtc_device {
 	enum rtc_wait_alarm_status wait_alarm_return_status;
 	struct stm32_exti_pdata *exti;
 	struct rtc *rtc;
-	bool alarm_wake;
 };
 
 /* Expect a single RTC instance */
@@ -1252,33 +1253,23 @@ stm32_rtc_wait_alarm(struct rtc *rtc __unused,
 	return TEE_SUCCESS;
 }
 
+#ifdef CFG_STM32_EXTI
 static TEE_Result stm32_rtc_set_alarm_wakeup_status(struct rtc *rtc __unused,
 						    bool status)
 {
 	if (!rtc_dev.rtc->is_wakeup_source)
 		return TEE_ERROR_NOT_SUPPORTED;
 
-	rtc_dev.alarm_wake = status;
-	return TEE_SUCCESS;
-}
-
-static TEE_Result
-stm32_rtc_pm(enum pm_op op, uint32_t pm_hint __unused,
-	     const struct pm_callback_handle *pm_handle __unused)
-{
-	if (op == PM_OP_SUSPEND) {
-#ifdef STM32_EXTI
-		if (rtc_dev.alarm_wake)
-			stm32_exti_enable_wake(rtc_dev.exti,
-					       rtc_dev.compat.exti_line_nb);
-		else
-			stm32_exti_disable_wake(rtc_dev.exti,
-						rtc_dev.compat.exti_line_nb);
-#endif
-	}
+	if (status)
+		stm32_exti_enable_wake(rtc_dev.exti,
+				       rtc_dev.compat.exti_line_nb);
+	else
+		stm32_exti_disable_wake(rtc_dev.exti,
+					rtc_dev.compat.exti_line_nb);
 
 	return TEE_SUCCESS;
 }
+#endif
 
 static const struct rtc_ops stm32_rtc_ops = {
 	.get_time = stm32_rtc_get_time,
@@ -1288,7 +1279,9 @@ static const struct rtc_ops stm32_rtc_ops = {
 	.enable_alarm = stm32_rtc_enable_alarm,
 	.wait_alarm = stm32_rtc_wait_alarm,
 	.cancel_wait = stm32_rtc_cancel_wait_alarm,
+#ifdef CFG_STM32_EXTI
 	.set_alarm_wakeup_status = stm32_rtc_set_alarm_wakeup_status,
+#endif
 };
 
 static struct rtc stm32_rtc = {
@@ -1366,10 +1359,6 @@ static TEE_Result stm32_rtc_probe(const void *fdt, int node,
 		if (res)
 			return res;
 
-		if (rtc_dev.rtc->is_wakeup_source)
-			register_pm_core_service_cb(stm32_rtc_pm, NULL,
-						    "stm32-rtc");
-
 		res = stm32_rtc_init();
 		if (res)
 			return res;
-- 
2.47.2

